### PR TITLE
Update Jest and test setup for `@testing-library/react`

### PR DESCRIPTION
**Pull request by joshblack**
This PR adds some quality of life improvements to testing along with removing some extraneous imports that could be consolidated into the test setup. Specifically, this PR does the following:

- Update Jest to v29
- Update `@storybook/test-runner` to v0.7.x
- Update snapshots to the new snapshot format in Jest v28 (brought in through Jest v29)
- Explicitly includes `jest-environment-jsdom` (through Jest v28 migration)
- Add `jest-watch-typeahead` watch plugin
- Removes explicit calls to `'babel-polyfill'` from test files
- Removes manual cleanup calls from `@testing-library/react` (this is done automatically in newer versions, I believe v12 is included in that)
- Removes calls to `@testing-library/jest-dom` (this is brought in through setup files)

**Remove `@testing-library/jest-dom` from individual test files**

Some test files were explicitly including `jest-dom` and were extending the test setup. This is being done in a setup file and so can be removed.

<details>
<summary>Files</summary>

- [x] `src/__tests__/Radio.test.tsx`
- [x] `src/__tests__/Overlay.test.tsx`
- [x] `src/__tests__/Checkbox.test.tsx`
- [x] `src/__tests__/Textarea.test.tsx`
- [x] `src/__tests__/ActionMenu.test.tsx`
- [x] `src/__tests__/Select.test.tsx`
- [x] `src/__tests__/ActionList.test.tsx`
</details>

**Remove `babel-polyfill` imports from individual test files**

`babel-polyfill`, and `@babel/polyfill`, [have been deprecated](https://babeljs.io/docs/en/babel-polyfill/) in favor of core-js. They can also be brought in through the setup test environment once instead of per-file.

<details>
<summary>Files</summary>

- [x] `src/__tests__/AnchoredOverlay.test.tsx`
- [x] `src/__tests__/BorderBox.test.tsx`
- [x] `src/__tests__/Flex.test.tsx`
- [x] `src/__tests__/FilterList.test.tsx`
- [x] `src/__tests__/LabelGroup.test.tsx`
- [x] `src/__tests__/Checkbox.test.tsx`
- [x] `src/__tests__/deprecated/ChoiceInputField.test.tsx`
- [x] `src/__tests__/deprecated/FormGroup.test.tsx`
- [x] `src/__tests__/deprecated/InputField.test.tsx`
- [x] `src/__tests__/deprecated/ActionList.test.tsx`
- [x] `src/__tests__/Tooltip.test.tsx`
- [x] `src/__tests__/deprecated/DropdownMenu.test.tsx`
- [x] `src/__tests__/Label.test.tsx`
- [x] `src/__tests__/deprecated/ActionMenu.test.tsx`
- [x] `src/__tests__/Flash.test.tsx`
- [x] `src/__tests__/Position.test.tsx`
- [x] `src/__tests__/deprecated/Label.test.tsx`
- [x] `src/__tests__/Avatar.test.tsx`
- [x] `src/__tests__/deprecated/Button.test.tsx`
- [x] `src/__tests__/TextInputWithTokens.test.tsx`
- [x] `src/__tests__/Caret.test.tsx`
- [x] `src/__tests__/Details.test.tsx`
- [x] `src/__tests__/Text.test.tsx`
- [x] `src/__tests__/Autocomplete.test.tsx`
- [x] `src/__tests__/Radio.test.tsx`
- [x] `src/__tests__/PointerBox.test.tsx`
- [x] `src/__tests__/BranchName.test.tsx`
- [x] `src/__tests__/UnderlineNav.test.tsx`
- [x] `src/__tests__/ProgressBar.test.tsx`
- [x] `src/__tests__/SelectPanel.test.tsx`
- [x] `src/__tests__/Header.test.tsx`
- [x] `src/__tests__/Dropdown.test.tsx`
- [x] `src/__tests__/FormControl.test.tsx`
- [x] `src/__tests__/StateLabel.test.tsx`
- [x] `src/__tests__/FilterListItem.test.tsx`
- [x] `src/__tests__/Textarea.test.tsx`
- [x] `src/__tests__/CircleBadge.test.tsx`
- [x] `src/__tests__/TextInput.test.tsx`
- [x] `src/__tests__/SideNav.test.tsx`
- [x] `src/__tests__/Select.test.tsx`
- [x] `src/__tests__/TabNav.test.tsx`
- [x] `src/__tests__/FilteredSearch.test.tsx`
- [x] `src/__tests__/utils/createSlots.test.tsx`
- [x] `src/__tests__/Heading.test.tsx`
- [x] `src/__tests__/ActionList.test.tsx`
- [x] `src/__tests__/ActionMenu.test.tsx`
- [x] `src/__tests__/Breadcrumbs.test.tsx`
- [x] `src/__tests__/UnderlineNavLink.test.tsx`
- [x] `src/__tests__/Link.test.tsx`
- [x] `src/__tests__/CounterLabel.test.tsx`
- [x] `src/__tests__/Token.test.tsx`
- [x] `src/__tests__/Popover.test.tsx`
- [x] `src/__tests__/Timeline.test.tsx`
- [x] `src/__tests__/Box.test.tsx`
- [x] `src/__tests__/Truncate.test.tsx`
- [x] `src/__tests__/Pagehead.test.tsx`
- [x] `src/__tests__/AvatarStack.test.tsx`
- [x] `src/__tests__/Pagination/Pagination.test.tsx`
- [x] `src/__tests__/Dialog.test.tsx`
- [x] `src/__tests__/SelectMenu.test.tsx`
- [x] `src/__tests__/BreadcrumbsItem.test.tsx`
- [x] `src/__tests__/CircleOcticon.test.tsx`
- [x] `src/__tests__/ConfirmationDialog.test.tsx`
- [x] `src/__tests__/Spinner.test.tsx`
- [x] `src/__tests__/Grid.test.tsx`
- [x] `src/__tests__/Pagination/PaginationModel.test.tsx`
- [x] `src/__tests__/SubNav.test.tsx`
- [x] `src/__tests__/StyledOcticon.test.tsx`
- [x] `src/__tests__/SubNavLink.test.tsx`
- [x] `src/__tests__/Button.test.tsx`
</details>

**Remove cleanup calls from individual test files**

I believe that `@testing-library/react` automatically calls `cleanup` if the test environment supports it. As a result, explicit imports and calls to `cleanup` have been removed in the following files.

<details>
<summary>Files</summary>

- [x] `src/__tests__/SubNavLink.test.tsx`
- [x] `src/__tests__/SideNav.test.tsx`
- [x] `src/__tests__/Tooltip.test.tsx`
- [x] `src/__tests__/Position.test.tsx`
- [x] `src/__tests__/Link.test.tsx`
- [x] `src/__tests__/Flash.test.tsx`
- [x] `src/__tests__/CircleOcticon.test.tsx`
- [x] `src/__tests__/Pagehead.test.tsx`
- [x] `src/__tests__/deprecated/FormGroup.test.tsx`
- [x] `src/__tests__/deprecated/Button.test.tsx`
- [x] `src/__tests__/UnderlineNav.test.tsx`
- [x] `src/__tests__/LabelGroup.test.tsx`
- [x] `src/__tests__/Details.test.tsx`
- [x] `src/__tests__/Caret.test.tsx`
- [x] `src/__tests__/BranchName.test.tsx`
- [x] `src/__tests__/Text.test.tsx`
- [x] `src/__tests__/Avatar.test.tsx`
- [x] `src/__tests__/Dropdown.test.tsx`
- [x] `src/__tests__/FilterList.test.tsx`
- [x] `src/__tests__/PointerBox.test.tsx`
- [x] `src/__tests__/StateLabel.test.tsx`
- [x] `src/__tests__/deprecated/Label.test.tsx`
- [x] `src/__tests__/UnderlineNavLink.test.tsx`
- [x] `src/__tests__/FilterListItem.test.tsx`
- [x] `src/__tests__/ProgressBar.test.tsx`
- [x] `src/__tests__/Heading.test.tsx`
- [x] `src/__tests__/FilteredSearch.test.tsx`
- [x] `src/__tests__/Breadcrumbs.test.tsx`
- [x] `src/__tests__/CircleBadge.test.tsx`
- [x] `src/__tests__/Popover.test.tsx`
- [x] `src/__tests__/Header.test.tsx`
- [x] `src/__tests__/Grid.test.tsx`
- [x] `src/__tests__/CounterLabel.test.tsx`
- [x] `src/__tests__/Timeline.test.tsx`
- [x] `src/__tests__/Spinner.test.tsx`
- [x] `src/__tests__/BorderBox.test.tsx`
- [x] `src/__tests__/AvatarStack.test.tsx`
- [x] `src/__tests__/Flex.test.tsx`
- [x] `src/__tests__/SubNav.test.tsx`
- [x] `src/__tests__/Truncate.test.tsx`
- [x] `src/__tests__/SelectMenu.test.tsx`
- [x] `src/__tests__/StyledOcticon.test.tsx`
- [x] `src/__tests__/Pagination/Pagination.test.tsx`

</details>

### CODE REVIEW

File: src/__tests__/ActionList.test.tsx [hunk 1]
```diff
@@ -1,12 +1,11 @@
-import {cleanup, render as HTMLRender, waitFor, fireEvent} from '@testing-library/react'
-import 'babel-polyfill'
+import {render as HTMLRender, waitFor, fireEvent} from '@testing-library/react'
+
```
Comment (joshblack) [inline src/__tests__/ActionList.test.tsx:2]:
```suggestion
```

File: src/__tests__/ActionMenu.test.tsx [hunk 2]
```diff
@@ -1,13 +1,12 @@
-import {cleanup, render as HTMLRender, waitFor} from '@testing-library/react'
+import {render as HTMLRender, waitFor} from '@testing-library/react'
 import userEvent from '@testing-library/user-event'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/ActionMenu.test.tsx:3]:
```suggestion
```

File: src/__tests__/AnchoredOverlay.test.tsx [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import React, {useCallback, useState} from 'react'
 import {AnchoredOverlay} from '../AnchoredOverlay'
 import {behavesAsComponent, checkExports} from '../utils/testing'
-import {render as HTMLRender, cleanup, fireEvent} from '@testing-library/react'
+import {render as HTMLRender, fireEvent} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/AnchoredOverlay.test.tsx:6]:
```suggestion
```

File: src/__tests__/Autocomplete.test.tsx [hunk 4]
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import {render} from '../utils/testing'
 import {render as HTMLRender, fireEvent, waitFor} from '@testing-library/react'
 import {toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/Autocomplete.test.tsx:5]:
```suggestion
```

File: src/__tests__/Box.test.tsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
-import {cleanup, render as HTMLRender} from '@testing-library/react'
-import 'babel-polyfill'
+import {render as HTMLRender} from '@testing-library/react'
+
```
Comment (joshblack) [inline src/__tests__/Box.test.tsx:2]:
```suggestion
```

File: src/__tests__/BreadcrumbsItem.test.tsx [hunk 6]
```diff
@@ -1,5 +1,5 @@
-import {cleanup, render as HTMLRender} from '@testing-library/react'
-import 'babel-polyfill'
+import {render as HTMLRender} from '@testing-library/react'
+
```
Comment (joshblack) [inline src/__tests__/BreadcrumbsItem.test.tsx:2]:
```suggestion
```

File: src/__tests__/Button.test.tsx [hunk 7]
```diff
@@ -1,9 +1,9 @@
 import React from 'react'
 import {IconButton, Button} from '../Button'
 import {behavesAsComponent} from '../utils/testing'
-import {render, cleanup, fireEvent} from '@testing-library/react'
+import {render, fireEvent} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/Button.test.tsx:6]:
```suggestion
```

File: src/__tests__/Checkbox.test.tsx [hunk 8]
```diff
@@ -1,18 +1,16 @@
 import React from 'react'
 import {Checkbox} from '..'
 import {behavesAsComponent, checkExports} from '../utils/testing'
-import {render, cleanup} from '@testing-library/react'
+import {render} from '@testing-library/react'
 import {toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
-import '@testing-library/jest-dom'
+
```
Comment (joshblack) [inline src/__tests__/Checkbox.test.tsx:6]:
```suggestion
```

File: src/__tests__/Dialog.test.tsx [hunk 9]
```diff
@@ -1,9 +1,9 @@
 import React, {useState, useRef} from 'react'
 import {Dialog, Box, Text} from '..'
 import {Button} from '../deprecated'
-import {render as HTMLRender, cleanup, act, fireEvent} from '@testing-library/react'
+import {render as HTMLRender, act, fireEvent} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/Dialog.test.tsx:6]:
```suggestion
```

File: src/__tests__/FormControl.test.tsx [hunk 10]
```diff
@@ -1,7 +1,7 @@
 import React from 'react'
-import {render, cleanup} from '@testing-library/react'
+import {render} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/FormControl.test.tsx:4]:
```suggestion
```

File: src/__tests__/Heading.test.tsx [hunk 11]
```diff
@@ -1,9 +1,9 @@
 import React from 'react'
 import {Heading} from '..'
 import {render, behavesAsComponent, checkExports} from '../utils/testing'
-import {render as HTMLRender, cleanup} from '@testing-library/react'
+import {render as HTMLRender} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/Heading.test.tsx:6]:
```suggestion
```

File: src/__tests__/Label.test.tsx [hunk 12]
```diff
@@ -1,7 +1,7 @@
 import React from 'react'
-import {render, cleanup} from '@testing-library/react'
+import {render} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/Label.test.tsx:4]:
```suggestion
```

File: src/__tests__/SelectMenu.test.tsx [hunk 13]
```diff
@@ -1,9 +1,9 @@
 import React from 'react'
 import {SelectMenu, Button} from '../deprecated'
 import {mount, render, renderRoot, COMPONENT_DISPLAY_NAME_REGEX, checkExports} from '../utils/testing'
-import {render as HTMLRender, cleanup} from '@testing-library/react'
+import {render as HTMLRender} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/SelectMenu.test.tsx:6]:
```suggestion
```

File: src/__tests__/SelectPanel.test.tsx [hunk 14]
```diff
@@ -1,5 +1,5 @@
-import {cleanup, render as HTMLRender} from '@testing-library/react'
-import 'babel-polyfill'
+import {render as HTMLRender} from '@testing-library/react'
+
```
Comment (joshblack) [inline src/__tests__/SelectPanel.test.tsx:2]:
```suggestion
```

File: src/__tests__/TabNav.test.tsx [hunk 15]
```diff
@@ -1,10 +1,10 @@
 import React from 'react'
 import {TabNav} from '..'
 import {mount, behavesAsComponent, checkExports} from '../utils/testing'
-import {fireEvent, render as HTMLRender, cleanup} from '@testing-library/react'
+import {fireEvent, render as HTMLRender} from '@testing-library/react'
 import userEvent from '@testing-library/user-event'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/TabNav.test.tsx:7]:
```suggestion
```

File: src/__tests__/TextInput.test.tsx [hunk 16]
```diff
@@ -1,9 +1,9 @@
 import React from 'react'
 import {TextInput} from '..'
 import {render, mount, behavesAsComponent, checkExports} from '../utils/testing'
-import {render as HTMLRender, cleanup, fireEvent} from '@testing-library/react'
+import {render as HTMLRender, fireEvent} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/TextInput.test.tsx:6]:
```suggestion
```

File: src/__tests__/TextInputWithTokens.test.tsx [hunk 17]
```diff
@@ -1,8 +1,8 @@
 import React from 'react'
 import {render} from '../utils/testing'
-import {render as HTMLRender, fireEvent, act, cleanup} from '@testing-library/react'
+import {render as HTMLRender, fireEvent, act} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/TextInputWithTokens.test.tsx:5]:
```suggestion
```

File: src/__tests__/Textarea.test.tsx [hunk 18]
```diff
@@ -1,10 +1,9 @@
 import React from 'react'
 import {Textarea} from '..'
 import {behavesAsComponent, checkExports, renderStyles} from '../utils/testing'
-import {render, cleanup} from '@testing-library/react'
+import {render} from '@testing-library/react'
 import {toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
-import '@testing-library/jest-dom'
+
```
Comment (joshblack) [inline src/__tests__/Textarea.test.tsx:6]:
```suggestion
```

File: src/__tests__/Token.test.tsx [hunk 19]
```diff
@@ -1,9 +1,9 @@
 import React from 'react'
 import Token from '../Token/Token'
 import {render, behavesAsComponent, mount} from '../utils/testing'
-import {render as HTMLRender, cleanup, fireEvent} from '@testing-library/react'
+import {render as HTMLRender, fireEvent} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/Token.test.tsx:6]:
```suggestion
```

File: src/__tests__/deprecated/ActionMenu.test.tsx [hunk 20]
```diff
@@ -1,5 +1,5 @@
-import {cleanup, render as HTMLRender, act, fireEvent} from '@testing-library/react'
-import 'babel-polyfill'
+import {render as HTMLRender, act, fireEvent} from '@testing-library/react'
+
```
Comment (joshblack) [inline src/__tests__/deprecated/ActionMenu.test.tsx:2]:
```suggestion
```

File: src/__tests__/deprecated/ActionList.test.tsx [hunk 21]
```diff
@@ -1,5 +1,5 @@
-import {cleanup, render as HTMLRender} from '@testing-library/react'
-import 'babel-polyfill'
+import {render as HTMLRender} from '@testing-library/react'
+
```
Comment (joshblack) [inline src/__tests__/deprecated/ActionList.test.tsx:2]:
```suggestion
```

File: src/__tests__/deprecated/ChoiceInputField.test.tsx [hunk 22]
```diff
@@ -1,7 +1,7 @@
 import React from 'react'
-import {render, cleanup} from '@testing-library/react'
+import {render} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/deprecated/ChoiceInputField.test.tsx:4]:
```suggestion
```

File: src/__tests__/deprecated/DropdownMenu.test.tsx [hunk 23]
```diff
@@ -1,5 +1,5 @@
-import {cleanup, render as HTMLRender, act, fireEvent} from '@testing-library/react'
-import 'babel-polyfill'
+import {render as HTMLRender, act, fireEvent} from '@testing-library/react'
+
```
Comment (joshblack) [inline src/__tests__/deprecated/DropdownMenu.test.tsx:2]:
```suggestion
```

File: src/__tests__/deprecated/InputField.test.tsx [hunk 24]
```diff
@@ -1,7 +1,7 @@
 import React from 'react'
-import {render, cleanup} from '@testing-library/react'
+import {render} from '@testing-library/react'
 import {axe, toHaveNoViolations} from 'jest-axe'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/deprecated/InputField.test.tsx:4]:
```suggestion
```

File: src/__tests__/utils/createSlots.test.tsx [hunk 25]
```diff
@@ -1,5 +1,5 @@
 import React from 'react'
-import 'babel-polyfill'
+
```
Comment (joshblack) [inline src/__tests__/utils/createSlots.test.tsx:2]:
```suggestion
```

**Review by colebemis — APPROVED** [review]
Thank you ✨

**Pull request merged**